<launch>

  <!-- KINEMATIC -->

  <!-- Robot kinematics class: one of 'TWO_WHEELED', 'HOLONOMIC'  -->
  <arg name="type" default="TWO_WHEELED"/>
  <!-- The radius -->
  <arg name="radius" default="0.3"/>
  <!-- The maximal speed -->
  <arg name="maximal_speed" default="0.3"/>
  <!-- The maximal angular speed -->
  <arg name="maximal_angular_speed" default=""/>

  <!-- KINEMATIC: TWO_WHEELED  -->
  <!-- The wheel axis length -->
  <arg name="axis_length" value="1.0"/>
  <!-- Unless maximal_angular_speed is defined: the maximal wheel speed while rotating,
   if undefined, will use maximal_speed-->
  <arg name="maximal_rotation_speed" value=""/>

  <!-- KINEMATIC: HOLONOMIC -->
  <!-- Enable point towards target -->
  <arg name="point_toward_target" default="False"/>


  <!-- PLANNING -->
  <!-- The navigation behavior. One of "HL", "ORCA", "HRVO" -->
  <arg name="behavior" default="0.3"/>
  <!-- The horizontal speed at which a free agent should move -->
  <arg name="optimal_speed" default="0.3"/>
  <!-- The angular speed at which a free agent should move [radian/sec]-->
  <arg name="optimal_angular_speed" default="0.3"/>
  <!-- or alternatively, for differential driven robots, the optimal wheel speed while rotating [m/s]-->
  <arg name="optimal_rotation_speed" default="0.3"/>
  <!-- The relaxation constant on speed -->
  <arg name="tau" default="0.5"/>
  <!-- The relaxation constant on angular speed -->
  <arg name="rotation_tau" default="0.5"/>
  <!-- The (planning) horizon -->
  <arg name="horizon" default="1"/>
  <!-- The safety margin to add to the robot radius-->
  <arg name="safety_margin" default="0.1"/>
  <!-- Target tolerance (angle) -->
  <arg name="tol_angle" default="0.1"/>
  <!-- Target tolerance (distance) -->
  <arg name="tol_distance" default="0.2"/>
  <!-- Minimal speed (lower desired speed are set to zero)-->
  <arg name="minimal_speed" default="0.05"/>


  <!-- PLANNING: HL -->
  <!-- The caution, i.e. how many seconds to keep away from an obstacle-->
  <arg name="eta" default="0.5"/>
  <!-- The aperture (i.e., the range of steering angles)-->
  <arg name="aperture" default="3.14"/>
  <!-- The resolution (i.e., the number of directions to subdivide aperture)-->
  <arg name="resolution" default="30"/>

  <!-- PLANNING: ORCA - HRVO -->
  <!-- The (planning) time horizon-->
  <arg name="time_horizon" default="1"/>

  <!-- PLANNING: 3D -->
  <!-- The vertical speed at which a free agent should move -->
  <arg name="optimal_vertical_speed" default="0.1"/>
  <!-- The relaxation constant on vertical velocity -->
  <arg name="tau_z" default="1"/>

  <!-- DEBUGGING -->
  <!-- enable (rviz) markers  -->
  <arg name="drawing" default='False'/>

  <!-- ROS -->
  <!-- Enable publish a TwistStamped instead of a Twist -->
  <arg name="publish_cmd_stamped" default='False'/>
  <!-- The control rate -->
  <arg name="rate" default='10'/>


   <node name="navigation" pkg="hl_navigation" type="controller" respawn="false" output="screen">
      <param name="type" value="$(arg type)" />
      <param name="radius" value="$(arg radius)" />
      <param name="axis_length" value="$(arg axis_length)" />
      <param name="maximal_speed" value="$(arg maximal_speed)" />
      <param name="maximal_angular_speed" value="$(arg maximal_angular_speed)" />
      <param name="maximal_rotation_speed" value="$(arg maximal_rotation_speed)" />
      <param name="behavior" value="$(arg behavior)" />

      <param name="publish_cmd_stamped" value="$(arg publish_cmd_stamped)" />
      <param name="rate" value="$(arg rate)" />

      <param name="optimalSpeed" value="$(arg optimalSpeed)"/>
      <param name="optimal_angular_speed" value="$(arg optimal_angular_speed)"/>
      <param name="optimal_rotation_speed" value="$(arg optimal_rotation_speed)"/>
      <param name="tau" value="$(arg tau)"/>
      <param name="rotation_tau" value="$(arg rotation_tau)"/>
      <param name="horizon" value="$(arg horizon)"/>
      <param name="safety_margin" value="$(arg safety_margin)"/>
      <param name="tol_angle" value="$(arg tol_angle)"/>
      <param name="tol_distance" value="$(arg tol_distance)"/>
      <param name="minimal_speed" value="$(arg minimal_speed)"/>

      <param name="optimal_vertical_speed" value="$(arg optimal_vertical_speed)"/>
      <param name="tau_z" value="$(arg tau_z)"/>

      <param name="eta" value="$(arg eta)"/>
      <param name="aperture" value="$(arg aperture)"/>
      <param name="resolution" value="$(arg resolution)"/>

      <param name="time_horizon" value="$(arg time_horizon)"/>

      <param name="drawing" value="$(arg drawing)"/>
   </node>
</launch>
